### PR TITLE
Customize Max IOBuffer Size

### PR DESCRIPTION
**Pull request by sudheerv**
Allow callers explicitly configure and pass in the max buffer sizes
when asking for unknown sized memory chunks depending on the usecase
instead of relying on a common/global default config.

This addresses https://github.com/apache/trafficserver/issues/6850

Email to dev/users@

```

Removing the config proxy.config.io.max_buffer_size
Yahoo
/
Sent



Sudheer Vinukonda <sudheervinukonda@yahoo.com>
To:
Users
,
Dev

Thu, Jun 4 at 12:52 PM

The config `proxy.config.io.max_buffer_size` determines the default Max IOBuffer size that's used when the caller (e.g HostDB alloc, POST request buffer) asks for a chunk of memory via the `iobuffer_size_to_index()` or `buffer_size_to_index()` interfaces without explicitly passing in the max IOBuffer. 

This can cause issues such as https://github.com/apache/trafficserver/issues/6850 or worse memory corruption such as https://github.com/apache/trafficserver/issues/6849

After some discussion with Leif, we think we should let callers explicitly pass in the max buffer sizes when asking for unknown sized memory chunks (e.g 
proxy.config.http.max_post_size and a new config 

proxy.config.hostdb.max_buffer.size) and remove the hidden default max io config.


Please let us know if there are any concerns/comments.

```

### CODE REVIEW

**Comment by sudheerv** [comment]
[approve ci autest]

**Comment by maskit** [comment]
[approve ci autest]

Rerunning autest because the last run didn't check anything.
```
fatal: ambiguous argument '73dde53575fad238d6c7396564bbe9a4336d5344^...73dde53575fad238d6c7396564bbe9a4336d5344': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
No relevant files changed, skipping run
Setting status of 73dde53575fad238d6c7396564bbe9a4336d5344 to SUCCESS with url https://ci.trafficserver.apache.org/job/autest-github/13427/ and message: 'Build finished. '
Using context: Jenkins autest
Finished: SUCCESS
```

**Comment by sudheerv** [comment]
[approve ci autest]

**Comment by maskit** [comment]
[approve ci autest]

Autest was skipped again. Maybe it happens after force-push?

**Comment by sudheerv** [comment]
[approve ci ubuntu]

**Comment by sudheerv** [comment]
> [approve ci autest]
> 
> Autest was skipped again. Maybe it happens after force-push?

Oh, that's odd. And ubuntu build was stuck in triggered state for hours too!

**Comment by sudheerv** [comment]
[approve ci autest]

**Comment by sudheerv** [comment]
[approve ci autest]

**Comment by sudheerv** [comment]
[approve ci autest]

**Comment by sudheerv** [comment]
[approve ci autest]

**Comment by sudheerv** [comment]
[approve ci autest]

**Comment by sudheerv** [comment]
@SolidWallOfCode @masaori335 @bryancall @vmamidi 

Can one of you pls review this? I'm running this PR in our prod with a bumped up buffer for HostDB to prevent the assert from #6850

**Comment by masaori335** [comment]
Adding `proxy.config.hostdb.io.max_buffer_index` for #6850 looks reasonable, but other changes I have questions.

1. `iobuffer_size_to_index()` vs `buffer_size_to_index()`
I don't know why, but we have similar functions. IIUC, the difference is `iobuffer_size_to_index()` returns negative value if requested size is larger than `BUFFER_SIZE_FOR_INDEX(max)`. OTOH, `buffer_size_to_index()` returns `0` as `DEFAULT_BUFFER_BASE_SIZE` in such case.
So we don't need to change `buffer_size_to_index()` calls if our target is `iobuffer_size_to_index()`.

2. If we this go this way, we should get rid of the default value of `iobuffer_size_to_index()` to make sure the caller sets the value explicitly.
```
- extern int64_t iobuffer_size_to_index(int64_t size, int64_t max = max_iobuffer_size);
+ extern int64_t iobuffer_size_to_index(int64_t size, int64_t max);
```

3. Some `iobuffer_size_to_index()` calls has value from dedicated records.config, others has fixed value (`BUFFER_SIZE_INDEX_32K`). What's the policy of this difference?

**Comment by sudheerv** [comment]
> Adding `proxy.config.hostdb.io.max_buffer_index` for #6850 looks reasonable, but other changes I have questions.
> 
> 1. `iobuffer_size_to_index()` vs `buffer_size_to_index()`
>    I don't know why, but we have similar functions. IIUC, the difference is `iobuffer_size_to_index()` returns negative value if requested size is larger than `BUFFER_SIZE_FOR_INDEX(max)`. OTOH, `buffer_size_to_index()` returns `0` as `DEFAULT_BUFFER_BASE_SIZE` in such case.

[Sudheer ] No, the idea is to enforce the caller pass the explicit buffer size as opposed to the iobuf library determining silently which could be problematic as it's hard to find a common default value that satisfies all use cases.

>    So we don't need to change `buffer_size_to_index()` calls if our target is `iobuffer_size_to_index()`.
> 2. If we go this way, we should get rid of the default value of `iobuffer_size_to_index()` to make sure the caller sets the value explicitly.
> 
> ```
> - extern int64_t iobuffer_size_to_index(int64_t size, int64_t max = max_iobuffer_size);
> + extern int64_t iobuffer_size_to_index(int64_t size, int64_t max);
> ```
> 

[Sudheer] 

+1. 

I removed it in my local patch that I'm testing in prod, but, forgot to include that change in the PR. Will add it now.

> 1. Some `iobuffer_size_to_index()` calls has value from dedicated records.config, others has fixed value (`BUFFER_SIZE_INDEX_32K`). What's the policy of this difference?

[Sudheer] Yeah, ideally we'd like all the use cases to have a config. I skipped QUIC use cases alone for the time being to revisit in the future. Do you recommend adding it to that as well?

**Comment by sudheerv** [comment]
[approve ci autest]

**Comment by sudheerv** [comment]
[approve ci autest]

**Comment by masaori335** [comment]
> > 1. `iobuffer_size_to_index()` vs `buffer_size_to_index()`
[snip]
> [Sudheer ] No, the idea is to enforce the caller pass the explicit buffer size as opposed to the iobuf library determining silently which could be problematic as it's hard to find a common default value that satisfies all use cases.
>

I agree with it's hard to find a common default value that satisfies all use cases. It's rarely tuned.

[snip]
 
> > 1. Some `iobuffer_size_to_index()` calls has value from dedicated records.config, others has fixed value (`BUFFER_SIZE_INDEX_32K`). What's the policy of this difference?
> 
> [Sudheer] Yeah, ideally we'd like all the use cases to have a config. I skipped QUIC use cases alone for the time being to revisit in the future when they are being used in prod as we'd then have a clearer picture on whether we need this stuff configurable and if so how many configs are needed (seems like there's tons of cases in QUIC). Do you recommend adding it to that as well?

I didn't notice they're only in the QUIC land. IMO, almost all of them are good with the current default value ( `BUFFER_SIZE_INDEX_32K` ). /cc @maskit

I worried about we need to add a ton of configs, but it looks like we can break down `proxy.config.io.max_buffer_size` into only 4 configs except for QUIC. Because many MIOBuffer has hardcoded value.

File: mgmt/RecordsConfig.cc [hunk 1]
```diff
@@ -720,6 +720,14 @@ static const RecordElement RecordsConfig[] =
   //##############################################################################
   {RECT_CONFIG, "proxy.config.io.max_buffer_size", RECD_INT, "32768", RECU_NULL, RR_NULL, RECC_NULL, nullptr, RECA_NULL}
   ,
+  {RECT_CONFIG, "proxy.config.hostdb.io.max_buffer_index", RECD_INT, "8", RECU_NULL, RR_NULL, RECC_NULL, nullptr, RECA_NULL}
```
Comment (masaori335) [inline mgmt/RecordsConfig.cc:723]: It'd be great if we have docs about which index means what size.

File: mgmt/RecordsConfig.cc [hunk 2]
```diff
@@ -720,6 +720,14 @@ static const RecordElement RecordsConfig[] =
   //##############################################################################
   {RECT_CONFIG, "proxy.config.io.max_buffer_size", RECD_INT, "32768", RECU_NULL, RR_NULL, RECC_NULL, nullptr, RECA_NULL}
```
Comment (masaori335) [inline mgmt/RecordsConfig.cc:721]: Should we get rid of this config now? This doesn't work anymore by this PR, right?

Comment (sudheerv) [reply mgmt/RecordsConfig.cc:721]: OK, yeah, I think that makes sense. It ended up resulting in a lot more changes across the code base, but, it sounds reasonable that it's done together. Will update the PR shortly with the new changes.

File: mgmt/RecordsConfig.cc (hunk 1 above)
Comment (sudheerv) [reply mgmt/RecordsConfig.cc:723]: Yeah, makes sense. Will add docs.

**Comment by sudheerv** [comment]
[approve ci Clang-Analyzer]

**Comment by sudheerv** [comment]
[approve ci autest]

**Comment by sudheerv** [comment]
[approve ci clang-format]

**Comment by sudheerv** [comment]
[approve ci autest]

**Comment by sudheerv** [comment]
[approve ci FreeBSD]

**Comment by sudheerv** [comment]
[approve ci Clang-Analyzer]

**Comment by sudheerv** [comment]
[approve ci autest]

**Comment by sudheerv** [comment]
[approve ci clang-format]

**Comment by sudheerv** [comment]
[approve ci RAT]

**Comment by sudheerv** [comment]
[approve ci Clang-Analyzer]

**Comment by masaori335** [comment]
Looks good for me. I'll approve when commits are squashed, but reviews by other eyeballs are appreciated because of the size of changes.

**Comment by sudheerv** [comment]
> Looks good for me. I'll approve when commits are squashed, but reviews by other eyeballs are appreciated because of the size of changes.

Hmm, I was going to use squash and merge as otherwise it'd trigger another round of CI dance?

File: doc/admin-guide/files/records.config.en.rst [hunk 3]
```diff
@@ -4407,6 +4407,31 @@ Sockets
    platforms.  (Currently only linux).  IO buffers are allocated with the MADV_DONTDUMP
    with madvise() on linux platforms that support MADV_DONTDUMP.  Enabled by default.
 
+.. ts:cv:: CONFIG proxy.config.ssl.misc.io.max_buffer_index INT 8
+
+   Configures the max IOBuffer Block index used for various SSL Operations
+   such as Handshake or Protocol Probe. Default value is 8 which maps to a 32K buffer
```
Comment (maskit) [inline doc/admin-guide/files/records.config.en.rst:4413]: Do we have a table for indexes and sizes on the docs? There should be `define` or `enum` in the code, but it's not very user friendly.

Comment (sudheerv) [reply doc/admin-guide/files/records.config.en.rst:4413]: Yeah, good point, but, could I do it in a separate PR later? Need to figure out the right doc to put that in and also the formatting etc.

Comment (maskit) [reply doc/admin-guide/files/records.config.en.rst:4413]: Sure, I don't think it's a blocker.

**Comment by sudheerv** [comment]
This fixes #6850

**Comment by zwoop** [comment]
Cherry-picked to v9.0.x branch.

**Pull request merged**